relay.go

```diff
@@ -61,6 +65,7 @@ func NewRelay(ctx context.Context, h host.Host, opts ...RelayOpt) (*Relay, error
 		self:     h.ID(),
 		incoming: make(chan *Conn),
 		relays:   make(map[peer.ID]struct{}),
+		liveHops: make(map[peer.ID]map[peer.ID]struct{}),
```

Comment:
FYI, you can do this with a single map and no counter:

```go
type hop struct {
  from peer.ID
  to peer.ID
}

// ...

liveHops: make(map[hop]struct{})
```

Reply:
also with double the memory complexity

Reply:
technically, i'm not using the maps here. But I would like to have that info exposed somehow

Reply:
Damn go has efficient maps. I was expecting the extra map to have a lot of memory overhead. Digression: The memory complexity depends on how many nodes we have and how many hops we have per node. For a sane number of nodes (<=1000), the two map method is more efficient assuming 3+ relays per node. >= 10000 nodes, the two map method is less efficient up to ~40 relays per node.

> technically, i'm not using the maps here. But I would like to have that info exposed somehow

(technically, you *can* still do that with the one map method, it's just harder to lookup info about a specific node)

---

For fun:

```go
package main

import (
	"fmt"
	"runtime"
)

type hop struct {
	from, to string
}

func main() {
	things := make([]string, 10000)
	for i := range things {
		things[i] = fmt.Sprintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx%d", i)
	}
	n := 37
	maps(things, n)
	nomaps(things, n)
}

func maps(things []string, n int) {
	var before, after runtime.MemStats

	runtime.ReadMemStats(&before)

	m := make(map[string]map[string]struct{})

	for _, f := range things {
		mm := make(map[string]struct{})
		for _, t := range things[:n] {
			mm[t] = struct{}{}
		}
		m[f] = mm
	}
	runtime.ReadMemStats(&after)
	fmt.Println(after.Alloc - before.Alloc)
}

func nomaps(things []string, n int) {
	var before, after runtime.MemStats

	runtime.ReadMemStats(&before)

	m := make(map[hop]struct{})

	for _, f := range things {
		for _, t := range things[:n] {
			m[hop{f, t}] = struct{}{}
		}
	}
	runtime.ReadMemStats(&after)
	fmt.Println(after.Alloc - before.Alloc)
}
```